bottlecap/src/lifecycle/flush_control.rs

```diff
@@ -0,0 +1,89 @@
+use::std::time;
+use crate::config::FlushStrategy;
+use crate::lifecycle::invocation_times::InvocationTimes;
+use tracing::debug;
+
+const TWENTY_SECONDS: u64 = 20 * 1000;
+
+pub struct FlushControl {
+    pub last_flush: u64,
+    flush_strategy: FlushStrategy,
+    invocation_times: InvocationTimes,
+}
+
+// FlushControl is called at the end of every invocation and decides whether or not we should flush
+// The flushing logic is complex and depends on the flush strategy
+// 1. Default Strategy
+//   - Flush at the end of the first 20 invocations
+//   - We keep track of the last 20 invocations and calculate the frequency.
+//     - If the duration from the last invocation to the 20th is less than 2 minutes, switch to
+//     periodic flush every 20s
+//     - else, flush at the end of the invocation
+//  2. Periodic Strategy
+//      - User specifies the interval in milliseconds
+//  3. End strategy
+//      - Always flush at the end of the invocation
+impl FlushControl {
+    pub fn new(flush_strategy: FlushStrategy) -> FlushControl {
+        FlushControl { flush_strategy, last_flush: 0, invocation_times: InvocationTimes::new() }
+    }
+
+    pub fn should_flush(&mut self) -> bool {
+        let now = time::SystemTime::now().duration_since(time::UNIX_EPOCH).unwrap().as_secs();
+        self.invocation_times.add(now);
+        match &self.flush_strategy {
+            FlushStrategy::Default => {
+                if self.invocation_times.should_adapt_to_periodic(now) {
+                    let should_periodic_flush = self.should_periodic_flush(TWENTY_SECONDS);
+                    debug!("Adapting over to periodic flush strategy. should_periodic_flush: {}", should_periodic_flush);
+                    return should_periodic_flush;
+                }
+                debug!("Not enough invocations to adapt to periodic flush, flushing at the end of the invocation");
+                self.last_flush = now;
+                true
+            },
+            FlushStrategy::Periodically(periodic) => self.should_periodic_flush(periodic.interval),
+            FlushStrategy::End => true,
+        }
+    }
+
+    fn should_periodic_flush(&mut self, interval: u64) -> bool {
+        let now = time::SystemTime::now().duration_since(time::UNIX_EPOCH).unwrap().as_secs();
```

Comment:
Why seconds?

Reply:
I just find it easier to reason about. That's why I divide off the millis later in this method. We could do millis if you wanted. Not sure we need the precision in this case